File: main/backend/app/api/v1/consent.py
"""個人情報保護法 / GDPR 風 同意管理 API."""
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Request
from pydantic import BaseModel, Field
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth import get_current_user_optional, get_tenant_id
from app.database import get_db
from app.models.consent import ConsentRecord, ConsentTemplate, DataDeletionRequest
from app.schemas.common import APIResponse
from app.services import consent_engine, consent_seeds

router = APIRouter(prefix="/api/v1/consent", tags=["consent"])


# ---- schemas ----

class ConsentTemplateRead(BaseModel):
    id: UUID
    code: str
    version: int
    title: str
    body_md: str
    required_fields: list[str] = Field(default_factory=list)


class GrantBody(BaseModel):
    template_id: UUID
    scope: dict | None = None  # {face: true, ...}


class WithdrawBody(BaseModel):
    record_id: UUID


class DeletionRequestBody(BaseModel):
    scope: str = Field(default="all", pattern="^(all|face|personal|training)$")
    note: str | None = None


class ConsentRecordRead(BaseModel):
    id: UUID
    template_id: UUID
    template_code: str
    template_version: int
    granted_at: datetime
    withdrawn_at: datetime | None = None
    scope: dict


class DeletionRequestRead(BaseModel):
    id: UUID
    user_id: UUID
    requested_by: UUID
    requested_at: datetime
    scope: str
    status: str
    processed_at: datetime | None = None
    deletion_log: dict | None = None
    note: str | None = None


# ---- helpers ----

def _user_id(user: dict | None) -> UUID:
    if not user or not user.get("sub"):
        raise HTTPException(status_code=401, detail="authentication required")
    try:
        return UUID(user["sub"])
    except Exception:
        raise HTTPException(status_code=400, detail="invalid user id")


def _is_admin(user: dict | None) -> bool:
    role = (user or {}).get("role")
    return role in ("admin", "owner")


# ---- endpoints ----

@router.get("/templates", response_model=APIResponse[list[ConsentTemplateRead]])
async def list_templates(
    db: AsyncSession = Depends(get_db),
    tenant_id: str = Depends(get_tenant_id),
):
    res = await db.execute(
        select(ConsentTemplate).where(ConsentTemplate.tenant_id == tenant_id).order_by(ConsentTemplate.code, ConsentTemplate.version)
    )
    rows = res.scalars().all()
    data = [
        ConsentTemplateRead(
            id=t.id, code=t.code, version=t.version, title=t.title, body_md=t.body_md,
            required_fields=list(t.required_fields_jsonb or []),
        )
        for t in rows
    ]
    return APIResponse(data=data)


@router.post("/seed", response_model=APIResponse[dict])
async def seed_templates(
    db: AsyncSession = Depends(get_db),
    tenant_id: str = Depends(get_tenant_id),
):
    """4テンプレ + 生体情報 marking を idempotent 投入."""
    tmpls = await consent_seeds.ensure_consent_templates(db, tenant_id)
    markings = await consent_seeds.ensure_biometric_markings(db, tenant_id)
    return APIResponse(data={"templates": tmpls, "markings": markings})


@router.get("/required", response_model=APIResponse[list[ConsentTemplateRead]])
async def required_consents(
    db: AsyncSession = Depends(get_db),
    tenant_id: str = Depends(get_tenant_id),
    user: dict | None = Depends(get_current_user_optional),
):
    """ユーザに必要 (= 未取得 or 撤回済み) の同意テンプレ一覧."""
    if not user or not user.get("sub"):
        # 未ログインなら全テンプレを返す (LP / 仮ログイン用)
        res = await db.execute(
            select(ConsentTemplate).where(ConsentTemplate.tenant_id == tenant_id)
        )
        rows = res.scalars().all()
    else:
        user_id = UUID(user["sub"])
        recs = (await db.execute(
            select(ConsentRecord).where(
                ConsentRecord.tenant_id == tenant_id,
                ConsentRecord.user_id == user_id,
                ConsentRecord.withdrawn_at.is_(None),
            )
        )).scalars().all()
        granted_codes = {r.template_code for r in recs}

        all_tmpls = (await db.execute(
            select(ConsentTemplate).where(ConsentTemplate.tenant_id == tenant_id)
        )).scalars().all()
        # code 単位で最新 version のみ
        latest_by_code: dict[str, ConsentTemplate] = {}
        for t in all_tmpls:
            cur = latest_by_code.get(t.code)
            if cur is None or t.version > cur.version:
                latest_by_code[t.code] = t
        rows = [t for code, t in latest_by_code.items() if code not in granted_codes]

    data = [
        ConsentTemplateRead(
            id=t.id, code=t.code, version=t.version, title=t.title, body_md=t.body_md,
            required_fields=list(t.required_fields_jsonb or []),
        )
        for t in rows
    ]
    return APIResponse(data=data)


@router.post("/grant", response_model=APIResponse[ConsentRecordRead])
async def grant(
    body: GrantBody,
    request: Request,
    db: AsyncSession = Depends(get_db),
    tenant_id: str = Depends(get_tenant_id),
    user: dict | None = Depends(get_current_user_optional),
):
    user_id = _user_id(user)
    ip = request.client.host if request.client else None
    try:
        rec = await consent_engine.record_consent(
            db, UUID(tenant_id), user_id, body.template_id, body.scope, ip
        )
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    return APIResponse(data=_record_dict(rec))


@router.post("/withdraw", response_model=APIResponse[ConsentRecordRead])
async def withdraw(
    body: WithdrawBody,
    db: AsyncSession = Depends(get_db),
    user: dict | None = Depends(get_current_user_optional),
):
    user_id = _user_id(user)
    rec = await consent_engine.withdraw(db, body.record_id)
    if rec is None:
        raise HTTPException(status_code=404, detail="record not found")
    if rec.user_id != user_id and not _is_admin(user):
        raise HTTPException(status_code=403, detail="cannot withdraw other user's consent")
    return APIResponse(data=_record_dict(rec))


@router.get("/my-consents", response_model=APIResponse[list[ConsentRecordRead]])
async def my_consents(
    db: AsyncSession = Depends(get_db),
    tenant_id: str = Depends(get_tenant_id),
    user: dict | None = Depends(get_current_user_optional),
):
    user_id = _user_id(user)
    res = await db.execute(
        select(ConsentRecord).where(
            ConsentRecord.tenant_id == tenant_id,
            ConsentRecord.user_id == user_id,
        ).order_by(ConsentRecord.granted_at.desc())
    )
    rows = res.scalars().all()
    return APIResponse(data=[_record_dict(r) for r in rows])


@router.post("/deletion-request", response_model=APIResponse[DeletionRequestRead])
async def create_deletion_request(
    body: DeletionRequestBody,
    db: AsyncSession = Depends(get_db),
    tenant_id: str = Depends(get_tenant_id),
    user: dict | None = Depends(get_current_user_optional),
):
    user_id = _user_id(user)
    req = DataDeletionRequest(
        tenant_id=UUID(tenant_id),
        user_id=user_id,
        requested_by=user_id,
        scope=body.scope,
        status="pending",
        note=body.note,
    )
    db.add(req)
    await db.commit()
    await db.refresh(req)
    return APIResponse(data=_deletion_dict(req))


@router.get("/deletion-requests", response_model=APIResponse[list[DeletionRequestRead]])
async def list_deletion_requests(
    db: AsyncSession = Depends(get_db),
    tenant_id: str = Depends(get_tenant_id),
    user: dict | None = Depends(get_current_user_optional),
):
    if not _is_admin(user):
        raise HTTPException(status_code=403, detail="admin only")
    res = await db.execute(
        select(DataDeletionRequest).where(
            DataDeletionRequest.tenant_id == tenant_id
        ).order_by(DataDeletionRequest.requested_at.desc())
    )
    rows = res.scalars().all()
    return APIResponse(data=[_deletion_dict(r) for r in rows])


@router.post("/deletion-requests/{request_id}/process", response_model=APIResponse[DeletionRequestRead])
async def process_deletion_request(
    request_id: UUID,
    db: AsyncSession = Depends(get_db),
    user: dict | None = Depends(get_current_user_optional),
):
    if not _is_admin(user):
        raise HTTPException(status_code=403, detail="admin only")
    processor_id = _user_id(user)
    try:
        req = await consent_engine.process_deletion(db, request_id, processor_id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    return APIResponse(data=_deletion_dict(req))


@router.post("/deletion-requests/{request_id}/reject", response_model=APIResponse[DeletionRequestRead])
async def reject_deletion_request(
    request_id: UUID,
    db: AsyncSession = Depends(get_db),
    user: dict | None = Depends(get_current_user_optional),
):
    if not _is_admin(user):
        raise HTTPException(status_code=403, detail="admin only")
    res = await db.execute(
        select(DataDeletionRequest).where(DataDeletionRequest.id == request_id)
    )
    req = res.scalar_one_or_none()
    if not req:
        raise HTTPException(status_code=404, detail="request not found")
    req.status = "rejected"
    req.processor_id = _user_id(user)
    await db.commit()
    await db.refresh(req)
    return APIResponse(data=_deletion_dict(req))


# ---- serialize helpers ----

def _record_dict(r: ConsentRecord) -> ConsentRecordRead:
    return ConsentRecordRead(
        id=r.id,
        template_id=r.template_id,
        template_code=r.template_code,
        template_version=r.template_version,
        granted_at=r.granted_at,
        withdrawn_at=r.withdrawn_at,
        scope=dict(r.scope_jsonb or {}),
    )


def _deletion_dict(r: DataDeletionRequest) -> DeletionRequestRead:
    return DeletionRequestRead(
        id=r.id,
        user_id=r.user_id,
        requested_by=r.requested_by,
        requested_at=r.requested_at,
        scope=r.scope,
        status=r.status,
        processed_at=r.processed_at,
        deletion_log=dict(r.deletion_log_jsonb or {}) if r.deletion_log_jsonb else None,
        note=r.note,
    )

File: main/frontend/src/lib/consent.ts
/**
 * Consent management client.
 *
 * - getRequiredConsents: returns templates that the current user has NOT yet
 *   granted. UI calls this on PWA boot and forces ConsentDialog if non-empty.
 * - grant / withdraw / requestDeletion: thin wrappers on the consent API.
 *
 * Mock fallback: when NEXT_PUBLIC_API_URL is empty (demo mode) we keep an
 * in-memory granted set + localStorage flag so the dialog only shows once.
 */
import { authFetch } from "./auth"

const API_URL = process.env.NEXT_PUBLIC_API_URL || ""
const LOCAL_KEY = "aentro.consent.granted_v1"

export interface ConsentTemplate {
  id: string
  code: string
  version: number
  title: string
  body_md: string
  required_fields: string[]
}

export interface ConsentRecord {
  id: string
  template_id: string
  template_code: string
  template_version: number
  granted_at: string
  withdrawn_at: string | null
  scope: Record<string, boolean>
}

export interface DeletionRequest {
  id: string
  user_id: string
  requested_by: string
  requested_at: string
  scope: "all" | "face" | "personal" | "training"
  status: "pending" | "processing" | "done" | "rejected"
  processed_at: string | null
  deletion_log: Record<string, unknown> | null
  note: string | null
}

const MOCK_TEMPLATES: ConsentTemplate[] = [
  {
    id: "tmpl-face",
    code: "face",
    version: 1,
    title: "顔認証データの取得・利用について",
    body_md:
      "# 顔認証データの取り扱い\n\n出退勤打刻のため、あなたの顔特徴量を取得・保管します。顔写真そのものは保存しません。",
    required_fields: ["face"],
  },
  {
    id: "tmpl-gps",
    code: "gps",
    version: 1,
    title: "位置情報（GPS）の取得について",
    body_md:
      "# 位置情報の取り扱い\n\n打刻時に店舗ジオフェンス内にいることを確認するため、緯度経度を取得します。常時取得はしません。",
    required_fields: ["gps"],
  },
  {
    id: "tmpl-clock",
    code: "clock_retention",
    version: 1,
    title: "打刻記録の保管について",
    body_md:
      "# 打刻記録（出勤簿）の保管\n\n労働基準法 109 条に基づき、出退勤・休憩記録を3年間保管します。",
    required_fields: ["clock_retention"],
  },
  {
    id: "tmpl-interview",
    code: "interview",
    version: 1,
    title: "面談・評価記録の保管について",
    body_md:
      "# 面談・評価記録\n\n店長・SVとの面談メモ、評価スコア、AI要約を保管します。テナント外部への学習提供は行いません。",
    required_fields: ["interview"],
  },
]

function readMockGranted(): Set<string> {
  if (typeof window === "undefined") return new Set()
  try {
    const raw = localStorage.getItem(LOCAL_KEY)
    if (!raw) return new Set()
    const arr = JSON.parse(raw) as string[]
    return new Set(arr)
  } catch {
    return new Set()
  }
}

function writeMockGranted(set: Set<string>): void {
  if (typeof window === "undefined") return
  localStorage.setItem(LOCAL_KEY, JSON.stringify(Array.from(set)))
}

export async function getRequiredConsents(): Promise<ConsentTemplate[]> {
  if (!API_URL) {
    const granted = readMockGranted()
    return MOCK_TEMPLATES.filter((t) => !granted.has(t.code))
  }
  try {
    const res = await authFetch("/api/v1/consent/required")
    if (!res.ok) return []
    const json = await res.json()
    return (json.data || []) as ConsentTemplate[]
  } catch {
    return []
  }
}

export async function grantConsent(templateId: string, scope: Record<string, boolean>): Promise<boolean> {
  if (!API_URL) {
    const granted = readMockGranted()
    const tmpl = MOCK_TEMPLATES.find((t) => t.id === templateId)
    if (tmpl) granted.add(tmpl.code)
    Object.entries(scope).forEach(([k, v]) => {
      if (v) granted.add(k)
    })
    writeMockGranted(granted)
    return true
  }
  const res = await authFetch("/api/v1/consent/grant", {
    method: "POST",
    body: JSON.stringify({ template_id: templateId, scope }),
  })
  return res.ok
}

export async function withdrawConsent(recordId: string): Promise<boolean> {
  if (!API_URL) {
    return true
  }
  const res = await authFetch("/api/v1/consent/withdraw", {
    method: "POST",
    body: JSON.stringify({ record_id: recordId }),
  })
  return res.ok
}

export async function listMyConsents(): Promise<ConsentRecord[]> {
  if (!API_URL) {
    const granted = readMockGranted()
    return Array.from(granted).map((code, i) => ({
      id: `mock-${i}`,
      template_id: `tmpl-${code}`,
      template_code: code,
      template_version: 1,
      granted_at: new Date().toISOString(),
      withdrawn_at: null,
      scope: { [code]: true },
    }))
  }
  const res = await authFetch("/api/v1/consent/my-consents")
  if (!res.ok) return []
  const json = await res.json()
  return (json.data || []) as ConsentRecord[]
}

export async function requestDataDeletion(
  scope: "all" | "face" | "personal" | "training",
  note?: string,
): Promise<DeletionRequest | null> {
  if (!API_URL) {
    return {
      id: `mock-${Date.now()}`,
      user_id: "demo",
      requested_by: "demo",
      requested_at: new Date().toISOString(),
      scope,
      status: "pending",
      processed_at: null,
      deletion_log: null,
      note: note ?? null,
    }
  }
  const res = await authFetch("/api/v1/consent/deletion-request", {
    method: "POST",
    body: JSON.stringify({ scope, note }),
  })
  if (!res.ok) return null
  const json = await res.json()
  return json.data as DeletionRequest
}

export async function listDeletionRequests(): Promise<DeletionRequest[]> {
  if (!API_URL) return []
  const res = await authFetch("/api/v1/consent/deletion-requests")
  if (!res.ok) return []
  const json = await res.json()
  return (json.data || []) as DeletionRequest[]
}

export async function processDeletionRequest(id: string): Promise<DeletionRequest | null> {
  if (!API_URL) return null
  const res = await authFetch(`/api/v1/consent/deletion-requests/${id}/process`, {
    method: "POST",
  })
  if (!res.ok) return null
  const json = await res.json()
  return json.data as DeletionRequest
}

export async function rejectDeletionRequest(id: string): Promise<DeletionRequest | null> {
  if (!API_URL) return null
  const res = await authFetch(`/api/v1/consent/deletion-requests/${id}/reject`, {
    method: "POST",
  })
  if (!res.ok) return null
  const json = await res.json()
  return json.data as DeletionRequest
}
